### PR TITLE
Fix2 get constraint of indexed access

### LINKED ISSUE #15371: anything is assignable to T[P] where T extends null/undefined/void/object/never
### case 1
```typescript
export function test<T extends object, P extends keyof T>(
    whatever: T, name: P
): T[P] {
    const result = Math.random() > 0.5 ? whatever[name] : 'Hello!'; // <-- expected T[P] | string, actual T[P]
    return result; // <-- returning unsound result
}
```

### case 2
```typescript
export function oneOrAnother<A, B>(one: A, another: B): A | B {
    return Math.random() > 0.5 ? one : another;
}

export function test<T extends object, P extends keyof T>(
    whatever: T, name: P
): T[P] {
    const result = oneOrAnother(whatever[name], 'Hello!'); // <-- result is clearly string | T[P] as expected
    return result; // <-- expected a type error, actual returning string | T[P] instead of T[P] no problem
}
```

### LINKED ISSUE #17069: Unable to add numeric types that are not simply 'number'
I'm trying to rewrite the signature of `sum` from `core.ts` in our compiler (see #16823), but I'm getting this error:

```ts
export function sum<T extends Record<K, number>, K extends string>(array: T[], prop: K): number {
    let result = 0;
    for (const v of array) {
        result += v[prop];
     // ~~~~~~~~~~~~~~~~~
     // Operator '+=' cannot be applied to types 'number' and 'T[K]'.
    }
    return result;
}
```

Note that I can't declare `sum` as

```ts
export function sum<K extends string>(array: Record<K, number>[], prop: K): number
```

because then TypeScript would infer *all* of the properties in each element for `K`. So I do need another type argument `T` for the indirection.

For a simpler repro:

```ts
function foo<T extends number>(x: T) {
    return 0 + x;
    //     ~~~~~
    // Operator '+' cannot be applied to types '0' and 'T'.

}
```

Related is #15645.

### LINKED ISSUE #17166: T[keyof T] assignable to T[K]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
function f<T, K extends keyof T>(obj: T, k: K, value:T[K]) {
  for (let key in obj) {
    value = obj[key];
    break;    
  } 
  return value;
}

f({ x: 1, y: "" }, 'y', "1").charAt(0); //runtime error

```

**Expected behavior:**
Compiler error 
**Actual behavior:**
Runtime error

### LINKED ISSUE #19298: Instanceof left-hand-side check rejects generic index access types
Reported by @yangchristian in https://github.com/Microsoft/TypeScript/issues/2775#issuecomment-337463284

I repro @larsenwork's issue with the following:
```ts
type CreatedOrUpdatedAt =
    { createdAt: Date | string } |
    { updatedAt: Date | string };

const datePropComp = <T extends CreatedOrUpdatedAt, K extends keyof T>(a: T, b: T, field: K) => {
    // The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
    const dateA = a[field] instanceof Date ? a[field] : new Date(a[field]);
    const dateB = b[field] instanceof Date ? b[field] : new Date(b[field]);

    return dateA.getTime() - dateB.getTime();
};

datePropComp({ updatedAt: Date() }, { updatedAt: Date() }, 'updatedAt' )
```
[Playground link](https://www.typescriptlang.org/play/#src=type%20CreatedOrUpdatedAt%20%3D%0A%20%20%20%20%7B%20createdAt%3A%20Date%20%7C%20string%20%7D%20%7C%0A%20%20%20%20%7B%20updatedAt%3A%20Date%20%7C%20string%20%7D%3B%0A%0Aconst%20datePropComp%20%3D%20%3CT%20extends%20CreatedOrUpdatedAt%2C%20K%20extends%20keyof%20T%3E(a%3A%20T%2C%20b%3A%20T%2C%20field%3A%20K)%20%3D%3E%20%7B%0A%20%20%20%20%2F%2F%20The%20left-hand%20side%20of%20an%20'instanceof'%20expression%20must%20be%20of%20type%20'any'%2C%20an%20object%20type%20or%20a%20type%20parameter.%0A%20%20%20%20const%20dateA%20%3D%20a%5Bfield%5D%20instanceof%20Date%20%3F%20a%5Bfield%5D%20%3A%20new%20Date(a%5Bfield%5D)%3B%0A%20%20%20%20const%20dateB%20%3D%20b%5Bfield%5D%20instanceof%20Date%20%3F%20b%5Bfield%5D%20%3A%20new%20Date(b%5Bfield%5D)%3B%0A%0A%20%20%20%20return%20dateA.getTime()%20-%20dateB.getTime()%3B%0A%7D%3B%0A%0AdatePropComp(%7B%20updatedAt%3A%20Date()%20%7D%2C%20%7B%20updatedAt%3A%20Date()%20%7D%2C%20'updatedAt'%20))

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #17069 
Fixes #17166
Fixes #15371
Fixes #19298
Fixes some regressions in RWC. 

1. `T[K]` now correctly produces `number` when `K extends string, T extends Record<K, number>`.
2. `T[K]` no longer allows any type to be assigned to it when `T extends object, K extends keyof T`.

Previously both of these cases failed in `getConstraintOfIndexedAccessType` because both bases followed `K`'s base constraint to `string` and then incorrectly produced `any` for types (like `object`) with no string index signature. In (1), this produced an error in `checkBinaryLikeExpression`. In (2), this failed to produce an error in `checkTypeRelatedTo`.

The fix has some additional pieces. 

In order to fix (1), not only does the index type parameter need to be kept around, the base constraint of `T extends Record<K, number>` needs to be `Record<K, number>` not `{}` so that `Record<K, number>[K]` produces `number`. So I changed `computeBaseConstraint` to only chase `T extends Record<K, number>` back to `Record<K, number>`, not all the way to `{}`.

In order to fix (2), any index type whose base constraint is string, whether it came from `keyof T` or not, is going to (incorrectly) produce `any` when indexing a type that doesn't have a string index signature -- the test case specifically tries `object[string]`. So I added an early exit to `getConstraintOfIndexedAccessType` that returns `undefined` in this case.

(This is a redo of #17870 after it got merged early by mistake.)

**No longer fixes** #17847, but previous commits contained a fix for it.

### CODE REVIEW

**Comment by sandersn** [comment]
@ahejlsberg this undoes [a change you added in the fix for #17521](https://github.com/Microsoft/TypeScript/pull/17521/commits/caea4f3a50ece23a0d7a3adb20b2d6ee227e7e8a#diff-c3ed224e4daa84352f7f1abcd23e8ccaR5987), but doesn't break any tests. I looked at the commit and [the associated tests](https://github.com/Microsoft/TypeScript/pull/17521/commits/98f6761590c5eb1e8aa388268e7df957234dd00c) and the rest of the change looks like it handles types with index signatures, not mapped types. Regardless, you should take a look and then we can talk about the right way to fix this bug (#17069) at the same time as #17521's fix.

**Comment by weswigham** [comment]
This fixes some new, erroneous errors in this example from our RWC tests:
```ts
// @declaration: true
// @jsx: react
// @module: commonjs
// @target: es5
// @experimentalDecorators: true
// @filename: index.ts
declare function cloneDeep(x: any): any;

abstract class ClonableModel<ActualType, CreatableType> {
    constructor(objectToEncapsulate?: CreatableType) {
        if (objectToEncapsulate) {
            for (var key in objectToEncapsulate) {
                if (objectToEncapsulate.hasOwnProperty(key)) {
                    this[key] = objectToEncapsulate[key];
                }
            }
        }
    }

    clone(): ActualType {
        let newT = new (Object.getPrototypeOf(this).constructor)();
        for (var key in this) {
            if (this.hasOwnProperty(key)) {
                const thisProp = this[key];
                if (thisProp instanceof ClonableModel) {
                    newT[key] = thisProp.clone();
                } else {
                    newT[key] = cloneDeep(thisProp);
                }
            }
        }
        return newT;
    }
}

export default ClonableModel;
```

**Comment by DanielRosenwasser** [comment]
Can you look at #17166 and decide whether it is related to this issue?

**Comment by sandersn** [comment]
It's related but this fix doesn't cover it. It the no-constraint exit explicitly only applies to `T`s that have constraints, but whose constraints don't have a string index signature. Of course unconstrained type parameters definitely don't have a constraint with an index signature either. I changed the check slightly and it now catches the first bug.

I still need to look at the second bug in #17166, but it appears to be different.

**Comment by sandersn** [comment]
The second bug in #17166 is not a bug. It's by design that `A[K]` is the union of all types of `A`, for some object type `A`.

**Comment by sandersn** [comment]
@ahejlsberg mind taking a look at this? It sounds like people keep running into it.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8350,7 +8356,7 @@ namespace ts {
 
         // Transform an indexed access to a simpler form, if possible. Return the simpler form, or return
         // undefined if no transformation is possible.
-        function getTransformedIndexedAccessType(type: IndexedAccessType): Type {
+        function simplifyIndexedAccessType(type: IndexedAccessType): Type {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: `getSimplifiedIndexAccessType`

**Comment by aleksey-bykov** [comment]
could this pr break https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-357051029 ?

**Pull request merged**